router/handler_link/bind.go

```diff
@@ -0,0 +1,127 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+func NewBindHandlerFactory(c xgress.CtrlChannel, f *forwarder.Forwarder, fo *forwarder.Options, mr metrics.Registry) *bindHandlerFactory {
+	return &bindHandlerFactory{
+		ctrl:             c,
+		forwarder:        f,
+		forwarderOptions: fo,
+		metricsRegistry:  mr,
+	}
+}
+
+type bindHandlerFactory struct {
+	ctrl             xgress.CtrlChannel
+	forwarder        *forwarder.Forwarder
+	forwarderOptions *forwarder.Options
+	metricsRegistry  metrics.Registry
+}
+
+func (self *bindHandlerFactory) NewBindHandler(link xlink.Xlink, latency bool, listenerSide bool) channel.BindHandler {
+	return &bindHandler{
+		bindHandlerFactory: self,
+		xlink:              link,
+		trackLatency:       latency,
+		listenerSide:       listenerSide,
+	}
+}
+
+type bindHandler struct {
+	*bindHandlerFactory
+	xlink        xlink.Xlink
+	trackLatency bool
+	listenerSide bool
+}
+
+func (self *bindHandler) BindChannel(binding channel.Binding) error {
+	ch := binding.GetChannel()
+	if self.listenerSide {
+		if err := self.verifyLink(self.xlink, ch); err != nil {
+			return err
+		}
+	}
+
+	closeNotify := make(chan struct{})
+
+	binding.GetChannel().SetLogicalName("l/" + self.xlink.Id().Token)
+	binding.SetUserData(self.xlink.Id().Token)
+	binding.AddCloseHandler(newCloseHandler(self.xlink, self.ctrl, self.forwarder, closeNotify))
+	binding.AddErrorHandler(newErrorHandler(self.xlink, self.ctrl))
+	binding.AddTypedReceiveHandler(newPayloadHandler(self.xlink, self.forwarder))
+	binding.AddTypedReceiveHandler(newQueuingAckHandler(self.xlink, self.forwarder, closeNotify))
+	binding.AddTypedReceiveHandler(&latency.LatencyHandler{})
+	binding.AddTypedReceiveHandler(newControlHandler(self.xlink, self.forwarder, closeNotify))
+	binding.AddPeekHandler(metrics2.NewChannelPeekHandler(self.xlink.Id().Token, self.forwarder.MetricsRegistry()))
+	binding.AddPeekHandler(trace.NewChannelPeekHandler(self.xlink.Id().Token, ch, self.forwarder.TraceController(), trace.NewChannelSink(self.ctrl.Channel())))
+
+	if self.trackLatency {
+		latencyMetric := self.metricsRegistry.Histogram("link." + self.xlink.Id().Token + ".latency")
+		latencyTimeout := self.forwarderOptions.LatencyProbeTimeout
+		config := &latency.ProbeConfig{
+			Channel:  ch,
+			Interval: self.forwarderOptions.LatencyProbeInterval,
+			Timeout:  latencyTimeout,
+			ResultHandler: func(resultNanos int64) {
+				latencyMetric.Update(resultNanos)
+			},
+			TimeoutHandler: func() {
+				pfxlog.ContextLogger(ch.Label()).Errorf("latency timeout after [%s] on channel [%s]", latencyTimeout, ch.Label())
+			},
+			ExitHandler: func() {
+				latencyMetric.Dispose()
+			},
+		}
+
+		go latency.ProbeLatencyConfigurable(config)
+	}
+
+	return nil
+}
+
+func (self *bindHandler) verifyLink(l xlink.Xlink, ch channel.Channel) error {
+	var fingerprints []string
+	for _, cert := range ch.Certificates() {
+		fingerprint := fmt.Sprintf("%x", sha1.Sum(cert.Raw))
```

Comment:
you'd think this was just some function. does it make sense to put somewhere that is reusable?

Reply:
I checked and there is an nfpem.FingerprintFromX509. I'll use that

Reply:
fixed